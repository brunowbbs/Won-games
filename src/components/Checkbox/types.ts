import { InputHTMLAttributes } from "react";

export type CheckboxProps = {
  label?: string;
  labelFor?: string;
  labelColor?: "white" | "black";
  isChecked?: boolean;
  onCheck?: (status: boolean) => void;
  value?: string | ReadonlyArray<string> | number;
} & InputHTMLAttributes<HTMLInputElement>;
