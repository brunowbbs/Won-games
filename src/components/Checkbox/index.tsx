import { useState } from "react";

import * as S from "./styles";
import { CheckboxProps } from "./types";

const Checkbox = ({
  label,
  labelFor = "",
  labelColor = "white",
  isChecked = false,
  onCheck,
  value,
  ...props
}: CheckboxProps) => {
  const [checked, setChecked] = useState(isChecked);

  const onChange = () => {
    const status = !checked;
    setChecked(status);

    onCheck && onCheck(status);
  };

  return (
    <S.Wrapper>
      <S.Input
        type="checkbox"
        id={labelFor}
        role="checkbox"
        onChange={onChange}
        checked={checked}
        value={value}
        {...props}
      />
      {!!label && (
        <S.Label htmlFor={labelFor} labelColor={labelColor}>
          {label}
        </S.Label>
      )}
    </S.Wrapper>
  );
};

export default Checkbox;
