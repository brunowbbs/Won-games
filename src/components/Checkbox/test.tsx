import { screen, render, waitFor, fireEvent } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { renderWithTheme } from "../../utils/tests/helpers";
import Checkbox from ".";

describe("<Checkbox", () => {
  it("Should render with label", () => {
    renderWithTheme(<Checkbox label="checkbox label" labelFor="check" />);

    expect(
      screen.getByRole("checkbox", { name: /checkbox label/i })
    ).toBeInTheDocument();

    expect(screen.getByLabelText(/checkbox label/i)).toBeInTheDocument();

    expect(screen.getByText(/checkbox label/i)).toHaveAttribute("for", "check");
  });

  it("Should render without label", () => {
    renderWithTheme(<Checkbox />);
    expect(screen.queryByLabelText("Checkbox")).not.toBeInTheDocument();
  });

  it("Should render with black label", () => {
    renderWithTheme(
      <Checkbox label="checkbox label" labelFor="check" labelColor="black" />
    );
    expect(screen.getByText(/checkbox label/i)).toHaveStyle({
      color: "#030517",
    });
  });

  it("Should dispatch onCheck when status change", async () => {
    const onCheck = jest.fn();
    renderWithTheme(<Checkbox label="Checkbox" onCheck={onCheck} />);
    expect(onCheck).not.toHaveBeenCalled();

    userEvent.click(screen.getByRole("checkbox"));
    await waitFor(() => {
      expect(onCheck).toHaveBeenCalledTimes(1);
    });

    expect(onCheck).toHaveBeenCalledWith(true);
  });

  it("Should dispatch onCheck when status change", async () => {
    const onCheck = jest.fn();
    renderWithTheme(<Checkbox label="Checkbox" onCheck={onCheck} isChecked />);
    expect(onCheck).not.toHaveBeenCalled();

    userEvent.click(screen.getByRole("checkbox"));
    await waitFor(() => {
      expect(onCheck).toHaveBeenCalledTimes(1);
    });

    expect(onCheck).toHaveBeenCalledWith(false);
  });

  it("should be accessible with tab", () => {
    renderWithTheme(<Checkbox label="Checkbox" labelFor="Checkbox" />);

    expect(document.body).toHaveFocus();

    userEvent.tab();
    expect(document.body).not.toHaveFocus();
    expect(screen.getByLabelText(/checkbox/i)).toHaveFocus();
  });
});
