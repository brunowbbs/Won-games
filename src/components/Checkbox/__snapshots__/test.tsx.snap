// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Checkbox Should render with label 1`] = `
<div
  class="sc-bcXHqe fMfVNB"
>
  <input
    class="sc-gswNZR bnjVLg"
    id="check"
    role="checkbox"
    type="checkbox"
    value=""
  />
  <label
    class="sc-dkrFOg beFfgO"
    for="check"
  >
    checkbox label
  </label>
</div>
`;
